### PR TITLE
[rest] Added support for default values containing a list of configuration options

### PR DESCRIPTION
**Pull request by cweitkamp**
- Added support for default values containing a list of configuration options by introducing an `EnrichedConfigDescriptionDTO`

Related to #998

If a configuration description defines `multiple="true"` the default value will be interpreted as a list and returned in a new JSON property `defaultValues`. Basically we can nullify the `defaultValue` in the JSON response in this case but that could be a breaking change as most of the UIs probably expect that property. The `defaultValue` itself will be split into multiple values according to the same logic like in #999.

The response for my example...

```
        <config-description uri="system:ephemeris">
		<parameter name="dayset-weekend" type="text" required="true" multiple="true">
			<label>Weekend Days</label>
			<description>List of usual non workable days.</description>
			<default>SATURDAY,SUNDAY</default>
		</parameter>
	</config-description>
```

will look like this:

```
{
  "uri": "system:ephemeris",
  "parameters": [
    {
      "defaultValues": [
          "SATURDAY",
          "SUNDAY"
      ],
      "defaultValue": "SATURDAY,SUNDAY",
      "description": "List of usual non workable days.",
      "label": "Weekend Days",
      "name": "dayset-weekend",
      "required": true,
      "type": "TEXT",
      "readOnly": false,
      "multiple": true,
      "advanced": false,
      "verify": false,
      "limitToOptions": true,
...
```

WIP for now because I think I missed my latest commit. And an unit test.

Signed-off-by: Christoph Weitkamp <github@christophweitkamp.de>

### CODE REVIEW

**Comment by cweitkamp** [comment]
Added missing commit and unit test.

File: bundles/org.openhab.core.io.rest.core/src/main/java/org/eclipse/smarthome/io/rest/core/config/EnrichedConfigDescriptionDTO.java [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+import org.eclipse.smarthome.config.core.dto.ConfigDescriptionParameterDTO;
+import org.eclipse.smarthome.config.core.dto.ConfigDescriptionParameterGroupDTO;
+
+/**
+ * This is an enriched data transfer object that is used to serialize {@link ConfigDescription}s.
+ *
+ * @author Christoph Weitkamp - Initial contribution
+ */
+@NonNullByDefault
+public class EnrichedConfigDescriptionDTO extends ConfigDescriptionDTO {
```
Comment (kaikreuzer) [inline bundles/org.openhab.core.io.rest.core/src/main/java/org/eclipse/smarthome/io/rest/core/config/EnrichedConfigDescriptionDTO.java]: Hm, I must be missing something. What is the difference between the `EnrichedConfigDescriptionDTO`and the normal `ConfigDescriptionDTO `? Why do we need this additional class at all?

Comment (cweitkamp) [reply bundles/org.openhab.core.io.rest.core/src/main/java/org/eclipse/smarthome/io/rest/core/config/EnrichedConfigDescriptionDTO.java]: I introduced this class to have a clear hierarchy of DTOs and to use it for the response definition of the REST API endpoints.

But in general you are right. The important thing is the extended `EnrichedConfigDescriptionParameterDTO`.

Comment (kaikreuzer) [reply bundles/org.openhab.core.io.rest.core/src/main/java/org/eclipse/smarthome/io/rest/core/config/EnrichedConfigDescriptionDTO.java]: Then I'd prefer to remove this class.

Comment (cweitkamp) [reply bundles/org.openhab.core.io.rest.core/src/main/java/org/eclipse/smarthome/io/rest/core/config/EnrichedConfigDescriptionDTO.java]: Done.

File: bundles/org.openhab.core.io.rest.core/src/main/java/org/eclipse/smarthome/io/rest/core/config/EnrichedConfigDescriptionDTOMapper.java [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+     * @return enriched configuration description DTO object
+     */
+    public static ConfigDescriptionDTO map(ConfigDescription configDescription) {
+        List<ConfigDescriptionParameterGroupDTO> parameterGroups = mapParameterGroups(
+                configDescription.getParameterGroups());
+        List<ConfigDescriptionParameterDTO> parameters = mapEnrichedParameters(configDescription.getParameters());
+        return new ConfigDescriptionDTO(toDecodedString(configDescription.getUID()), parameters, parameterGroups);
+    }
+
+    public static List<ConfigDescriptionParameterDTO> mapEnrichedParameters(
```
Comment (kaikreuzer) [inline bundles/org.openhab.core.io.rest.core/src/main/java/org/eclipse/smarthome/io/rest/core/config/EnrichedConfigDescriptionDTOMapper.java]: Shouldn't this be a private method? If not, JavaDoc is missing 😉. 
```suggestion
    private static List<ConfigDescriptionParameterDTO> mapEnrichedParameters(
```

File: bundles/org.openhab.core.io.rest.core/src/main/java/org/eclipse/smarthome/io/rest/core/config/EnrichedConfigDescriptionParameterDTO.java [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+
+import org.eclipse.jdt.annotation.NonNullByDefault;
+import org.eclipse.jdt.annotation.Nullable;
+import org.eclipse.smarthome.config.core.ConfigDescriptionParameter.Type;
+import org.eclipse.smarthome.config.core.dto.ConfigDescriptionParameterDTO;
+import org.eclipse.smarthome.config.core.dto.FilterCriteriaDTO;
+import org.eclipse.smarthome.config.core.dto.ParameterOptionDTO;
+
+/**
+ * This is a data transfer object that is used to serialize parameter of an enriched configuration description.
```
Comment (kaikreuzer) [inline bundles/org.openhab.core.io.rest.core/src/main/java/org/eclipse/smarthome/io/rest/core/config/EnrichedConfigDescriptionParameterDTO.java]: Could you add a sentence about what this enriched version of the DTO is (what's the difference to the "normal" DTO)?

File: bundles/org.openhab.core.io.rest.core/src/main/java/org/eclipse/smarthome/io/rest/core/config/EnrichedConfigDescriptionDTOMapper.java (hunk 2 above)
Comment (cweitkamp) [reply bundles/org.openhab.core.io.rest.core/src/main/java/org/eclipse/smarthome/io/rest/core/config/EnrichedConfigDescriptionDTOMapper.java]: I added JavaDoc as the `mapParameters()` method of the base class is public too.

File: bundles/org.openhab.core.io.rest.core/src/main/java/org/eclipse/smarthome/io/rest/core/config/EnrichedConfigDescriptionParameterDTO.java (hunk 3 above)
Comment (cweitkamp) [reply bundles/org.openhab.core.io.rest.core/src/main/java/org/eclipse/smarthome/io/rest/core/config/EnrichedConfigDescriptionParameterDTO.java]: Done.

**Comment by charly61ch** [comment]
**Z-Wave** Side effect !?

Dears,
Since modification of the this file (ConfigDescriptionResource.java), I observe a bug in the configuration of **at least** one ZW-device binding.
I don't have a lot of devices, just a "any" z-wave device for which I don't observe the problem, and 
a **Horstman hrt4zw** thermostat (hrt4zw_0_0.xml), for which  I observe the following :
- I include the device : no problem.
- I (auto) link the channels, and I see them in the paper-ui's control page, perfectly communicating : no problem.
- In the "Configuration -> Things" page of this device, I can link / unlink channels, see its properties : no problem.

**BUT (the problem)** :  as soon I press the "pencil" icon, for modifiyng the parameters, I receive on my web page, the error 500, server error.

I use the latest snapshot for both z-wave and core (about 17~18 oct. ).
If I build the core at the snapshot of today (19. oct.) but with the only "ConfigDescriptionResource.java" just before the commit 1031, all is OK.

I think there is a problem behind this behavior, which shows an unsupported / mistake in zwave binding.

I hope that my observation serves.
Regards,
C.Cléments

**Comment by wborn** [comment]
Do you see the same NPE as reported in the [community thread](https://community.openhab.org/t/snapshot-1728-no-advanced-settings-in-habmin-error-500-in-paper-ui-for-zwave-things/83789?u=wborn) @charly61ch when raising the Jersey log level with:

`log:set trace org.glassfish.jersey.server`

Logging: 

```
Caused by: java.lang.NullPointerException
	at org.eclipse.smarthome.io.rest.core.config.EnrichedConfigDescriptionParameterDTO.<init>(EnrichedConfigDescriptionParameterDTO.java:52) ~[?:?]
	at org.eclipse.smarthome.io.rest.core.config.EnrichedConfigDescriptionDTOMapper.mapEnrichedParameters(EnrichedConfigDescriptionDTOMapper.java:59) ~[?:?]
	at org.eclipse.smarthome.io.rest.core.config.EnrichedConfigDescriptionDTOMapper.map(EnrichedConfigDescriptionDTOMapper.java:45) ~[?:?]
	at org.eclipse.smarthome.io.rest.core.internal.config.ConfigDescriptionResource.getByURI(ConfigDescriptionResource.java:110) ~[?:?]
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[?:?]
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62) ~[?:?]
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[?:?]
	at java.lang.reflect.Method.invoke(Method.java:498) ~[?:?]
	at org.glassfish.jersey.server.model.internal.ResourceMethodInvocationHandlerFactory$1.invoke(ResourceMethodInvocationHandlerFactory.java:81) ~[?:?]
	at org.glassfish.jersey.server.model.internal.AbstractJavaResourceMethodDispatcher$1.run(AbstractJavaResourceMethodDispatcher.java:144) ~[?:?]
	... 54 more
```

Can you have a look @cweitkamp?

**Comment by cweitkamp** [comment]
@born Sure. Should be no big deal.

As a side note: I started working on nullness annotations for `org.openhab.core.config.core` during this implementation. Will continue on it now ...

**Comment by charly61ch** [comment]
@wborn @cweitkamp 
Yes, :
(note also the 1st line,.... not sure, but it seems I never seen (not sure !?)

2019-10-19 15:25:58.453 [INFO ] [ommandclass.ZWaveVersionCommandClass] - NODE 2: Command Class COMMAND_CLASS_BASIC has version 0!
2019-10-19 15:28:59.367 [DEBUG] [ersey.server.ServerRuntime$Responder] - Starting mapping of the exception.
org.glassfish.jersey.server.internal.process.MappableException: java.lang.NullPointerException
	at org.glassfish.jersey.server.model.internal.AbstractJavaResourceMethodDispatcher.mapTargetToRuntimeEx(AbstractJavaResourceMethodDispatcher.java:179) ~[125:org.glassfish.jersey.core.jersey-server:2.22.2]
	at org.glassfish.jersey.server.model.internal.AbstractJavaResourceMethodDispatcher.access$200(AbstractJavaResourceMethodDispatcher.java:72) ~[125:org.glassfish.jersey.core.jersey-server:2.22.2]
	at org.glassfish.jersey.server.model.internal.AbstractJavaResourceMethodDispatcher$1.run(AbstractJavaResourceMethodDispatcher.java:149) ~[125:org.glassfish.jersey.core.jersey-server:2.22.2]
	at org.glassfish.jersey.server.model.internal.AbstractJavaResourceMethodDispatcher.invoke(AbstractJavaResourceMethodDispatcher.java:161) ~[125:org.glassfish.jersey.core.jersey-server:2.22.2]
	at org.glassfish.jersey.server.model.internal.JavaResourceMethodDispatcherProvider$ResponseOutInvoker.doDispatch(JavaResourceMethodDispatcherProvider.java:160) ~[125:org.glassfish.jersey.core.jersey-server:2.22.2]
	at org.glassfish.jersey.server.model.internal.AbstractJavaResourceMethodDispatcher.dispatch(AbstractJavaResourceMethodDispatcher.java:99) ~[125:org.glassfish.jersey.core.jersey-server:2.22.2]
	at org.glassfish.jersey.server.model.ResourceMethodInvoker.invoke(ResourceMethodInvoker.java:389) ~[125:org.glassfish.jersey.core.jersey-server:2.22.2]
	at org.glassfish.jersey.server.model.ResourceMethodInvoker.apply(ResourceMethodInvoker.java:347) ~[125:org.glassfish.jersey.core.jersey-server:2.22.2]
	at org.glassfish.jersey.server.model.ResourceMethodInvoker.apply(ResourceMethodInvoker.java:102) ~[125:org.glassfish.jersey.core.jersey-server:2.22.2]
	at org.glassfish.jersey.server.ServerRuntime$2.run(ServerRuntime.java:326) [125:org.glassfish.jersey.core.jersey-server:2.22.2]
	at org.glassfish.jersey.internal.Errors$1.call(Errors.java:271) [124:org.glassfish.jersey.core.jersey-common:2.22.2]
	at org.glassfish.jersey.internal.Errors$1.call(Errors.java:267) [124:org.glassfish.jersey.core.jersey-common:2.22.2]
	at org.glassfish.jersey.internal.Errors.process(Errors.java:315) [124:org.glassfish.jersey.core.jersey-common:2.22.2]
	at org.glassfish.jersey.internal.Errors.process(Errors.java:297) [124:org.glassfish.jersey.core.jersey-common:2.22.2]
	at org.glassfish.jersey.internal.Errors.process(Errors.java:267) [124:org.glassfish.jersey.core.jersey-common:2.22.2]
	at org.glassfish.jersey.process.internal.RequestScope.runInScope(RequestScope.java:317) [124:org.glassfish.jersey.core.jersey-common:2.22.2]
	at org.glassfish.jersey.server.ServerRuntime.process(ServerRuntime.java:305) [125:org.glassfish.jersey.core.jersey-server:2.22.2]
	at org.glassfish.jersey.server.ApplicationHandler.handle(ApplicationHandler.java:1154) [125:org.glassfish.jersey.core.jersey-server:2.22.2]
	at org.glassfish.jersey.servlet.WebComponent.serviceImpl(WebComponent.java:473) [122:org.glassfish.jersey.containers.jersey-container-servlet-core:2.22.2]
	at org.glassfish.jersey.servlet.WebComponent.service(WebComponent.java:427) [122:org.glassfish.jersey.containers.jersey-container-servlet-core:2.22.2]
	at org.glassfish.jersey.servlet.ServletContainer.service(ServletContainer.java:388) [122:org.glassfish.jersey.containers.jersey-container-servlet-core:2.22.2]
	at org.glassfish.jersey.servlet.ServletContainer.service(ServletContainer.java:341) [122:org.glassfish.jersey.containers.jersey-container-servlet-core:2.22.2]
	at org.glassfish.jersey.servlet.ServletContainer.service(ServletContainer.java:228) [122:org.glassfish.jersey.containers.jersey-container-servlet-core:2.22.2]
	at com.eclipsesource.jaxrs.publisher.internal.ServletContainerBridge.service(ServletContainerBridge.java:76) [20:com.eclipsesource.jaxrs.publisher:5.3.1.201602281253]
	at org.eclipse.jetty.servlet.ServletHolder.handle(ServletHolder.java:873) [90:org.eclipse.jetty.servlet:9.4.18.v20190429]
	at org.eclipse.jetty.servlet.ServletHandler.doHandle(ServletHandler.java:542) [90:org.eclipse.jetty.servlet:9.4.18.v20190429]
	at org.ops4j.pax.web.service.jetty.internal.HttpServiceServletHandler.doHandle(HttpServiceServletHandler.java:71) [194:org.ops4j.pax.web.pax-web-jetty:7.2.10]
	at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:146) [89:org.eclipse.jetty.server:9.4.18.v20190429]
	at org.eclipse.jetty.security.SecurityHandler.handle(SecurityHandler.java:548) [87:org.eclipse.jetty.security:9.4.18.v20190429]
	at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:132) [89:org.eclipse.jetty.server:9.4.18.v20190429]
	at org.eclipse.jetty.server.handler.ScopedHandler.nextHandle(ScopedHandler.java:257) [89:org.eclipse.jetty.server:9.4.18.v20190429]
	at org.eclipse.jetty.server.session.SessionHandler.doHandle(SessionHandler.java:1700) [89:org.eclipse.jetty.server:9.4.18.v20190429]
	at org.eclipse.jetty.server.handler.ScopedHandler.nextHandle(ScopedHandler.java:255) [89:org.eclipse.jetty.server:9.4.18.v20190429]
	at org.eclipse.jetty.server.handler.ContextHandler.doHandle(ContextHandler.java:1345) [89:org.eclipse.jetty.server:9.4.18.v20190429]
	at org.ops4j.pax.web.service.jetty.internal.HttpServiceContext.doHandle(HttpServiceContext.java:293) [194:org.ops4j.pax.web.pax-web-jetty:7.2.10]
	at org.eclipse.jetty.server.handler.ScopedHandler.nextScope(ScopedHandler.java:203) [89:org.eclipse.jetty.server:9.4.18.v20190429]
	at org.eclipse.jetty.servlet.ServletHandler.doScope(ServletHandler.java:480) [90:org.eclipse.jetty.servlet:9.4.18.v20190429]
	at org.eclipse.jetty.server.session.SessionHandler.doScope(SessionHandler.java:1667) [89:org.eclipse.jetty.server:9.4.18.v20190429]
	at org.eclipse.jetty.server.handler.ScopedHandler.nextScope(ScopedHandler.java:201) [89:org.eclipse.jetty.server:9.4.18.v20190429]
	at org.eclipse.jetty.server.handler.ContextHandler.doScope(ContextHandler.java:1247) [89:org.eclipse.jetty.server:9.4.18.v20190429]
	at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:144) [89:org.eclipse.jetty.server:9.4.18.v20190429]
	at org.ops4j.pax.web.service.jetty.internal.JettyServerHandlerCollection.handle(JettyServerHandlerCollection.java:80) [194:org.ops4j.pax.web.pax-web-jetty:7.2.10]
	at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:132) [89:org.eclipse.jetty.server:9.4.18.v20190429]
	at org.eclipse.jetty.server.Server.handle(Server.java:505) [89:org.eclipse.jetty.server:9.4.18.v20190429]
	at org.eclipse.jetty.server.HttpChannel.handle(HttpChannel.java:370) [89:org.eclipse.jetty.server:9.4.18.v20190429]
	at org.eclipse.jetty.server.HttpConnection.onFillable(HttpConnection.java:267) [89:org.eclipse.jetty.server:9.4.18.v20190429]
	at org.eclipse.jetty.io.AbstractConnection$ReadCallback.succeeded(AbstractConnection.java:305) [80:org.eclipse.jetty.io:9.4.18.v20190429]
	at org.eclipse.jetty.io.FillInterest.fillable(FillInterest.java:103) [80:org.eclipse.jetty.io:9.4.18.v20190429]
	at org.eclipse.jetty.io.ChannelEndPoint$2.run(ChannelEndPoint.java:117) [80:org.eclipse.jetty.io:9.4.18.v20190429]
	at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.runTask(EatWhatYouKill.java:333) [92:org.eclipse.jetty.util:9.4.18.v20190429]
	at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.doProduce(EatWhatYouKill.java:310) [92:org.eclipse.jetty.util:9.4.18.v20190429]
	at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.tryProduce(EatWhatYouKill.java:168) [92:org.eclipse.jetty.util:9.4.18.v20190429]
	at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.run(EatWhatYouKill.java:126) [92:org.eclipse.jetty.util:9.4.18.v20190429]
	at org.eclipse.jetty.util.thread.ReservedThreadExecutor$ReservedThread.run(ReservedThreadExecutor.java:366) [92:org.eclipse.jetty.util:9.4.18.v20190429]
	at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:698) [92:org.eclipse.jetty.util:9.4.18.v20190429]
	at org.eclipse.jetty.util.thread.QueuedThreadPool$Runner.run(QueuedThreadPool.java:804) [92:org.eclipse.jetty.util:9.4.18.v20190429]
	at java.lang.Thread.run(Thread.java:748) [?:?]
Caused by: java.lang.NullPointerException
	at org.eclipse.smarthome.io.rest.core.config.EnrichedConfigDescriptionParameterDTO.<init>(EnrichedConfigDescriptionParameterDTO.java:52) ~[?:?]
	at org.eclipse.smarthome.io.rest.core.config.EnrichedConfigDescriptionDTOMapper.mapEnrichedParameters(EnrichedConfigDescriptionDTOMapper.java:59) ~[?:?]
	at org.eclipse.smarthome.io.rest.core.config.EnrichedConfigDescriptionDTOMapper.map(EnrichedConfigDescriptionDTOMapper.java:45) ~[?:?]
	at org.eclipse.smarthome.io.rest.core.internal.config.ConfigDescriptionResource.getByURI(ConfigDescriptionResource.java:110) ~[?:?]
	at sun.reflect.GeneratedMethodAccessor19.invoke(Unknown Source) ~[?:?]
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[?:?]
	at java.lang.reflect.Method.invoke(Method.java:498) ~[?:?]
	at org.glassfish.jersey.server.model.internal.ResourceMethodInvocationHandlerFactory$1.invoke(ResourceMethodInvocationHandlerFactory.java:81) ~[?:?]
	at org.glassfish.jersey.server.model.internal.AbstractJavaResourceMethodDispatcher$1.run(AbstractJavaResourceMethodDispatcher.java:144) ~[?:?]
	... 54 more
2019-10-19 15:28:59.380 [DEBUG] [ersey.server.ServerRuntime$Responder] - Exception 'null' has been mapped by 'org.eclipse.smarthome.io.rest.JSONResponse$ExceptionMapper' to response 'Internal Server Error' (500:SERVER_ERROR).

**Comment by cweitkamp** [comment]
Thanks for reporting and evaluating. I submitted a fix (see https://github.com/openhab/openhab-core/pull/1149).

**Pull request merged**